velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
     child->loadedVector();
   }
   input_ = std::move(input);
+  VELOX_CHECK_EQ(buildIndex_, 0);
+  if (needsProbeMismatch(joinType_)) {
+    probeMatched_.resizeFill(input_->size(), false);
+  }
 }
 
 RowVectorPtr NestedLoopJoinProbe::getOutput() {
-  if (input_ == nullptr) {
+  if (isFinished()) {
     return nullptr;
   }
+  RowVectorPtr output{nullptr};
+  while (output == nullptr) {
+    if (lastProbe_) {
+      VELOX_CHECK(processingBuildMismatch());
+      VELOX_CHECK_NULL(input_);
 
-  const auto inputSize = input_->size();
+      while (output == nullptr && !hasProbedAllBuildData()) {
+        output = getMismatchedOutput(
+            buildVectors_.value()[buildIndex_],
+            buildMatched_[buildIndex_],
+            buildOutMapping_,
+            buildProjections_,
+            identityProjections_);
+        ++buildIndex_;
+      }
+      if (hasProbedAllBuildData()) {
+        setState(ProbeOperatorState::kFinish);
+      }
+      break;
+    }
+
+    if (input_ == nullptr) {
+      break;
+    }
+
+    // When input_ is not null but buildIndex_ is at the end, it means the
+    // matching of input_ and buildData_ has finished. For left/full joins,
+    // the next step is to emit output for mismatched probe side rows.
+    if (hasProbedAllBuildData()) {
+      output = needsProbeMismatch(joinType_) ? getMismatchedOutput(
```

Comment:
It seems that needsProbeMismatch is always true in this case? Shall we add CHECK instead?

Reply:
> It seems that needsProbeMismatch is always true in this case?

The only case it's false I can think of is empty build data.